# . Introduction-The World Bank International Debt Data

It's not that we humans only take debts to manage our necessities. A country may also take debt to manage its economy. For example, infrastructure spending is one costly ingredient required for a country's citizens to lead comfortable lives. The World Bank is one of the the organization that provides debt to countries.

In this notebook, we are going to analyze international debt data collected by The World Bank. The dataset contains information about the amount of debt (in USD) owed by developing countries across several categories. We are going to find the answers to questions like:

- What is the total amount of debt that is owed by the countries listed in the dataset?
- Which country owns the maximum amount of debt and what does that amount look like?
- What is the average amount of debt owed by countries across different debt indicators?


Here we will be using SQL queries only to understand and analyze this dataset.

# . Importing the important libraries for analysis of the data.

In [ ]:
import sqlalchemy
import pyodbc
import pymysql

In [3]:
import pandas as pd
import numpy as np

In [4]:
#It is used for loading the magic cell sql commands.
%load_ext sql  

# . Connecting with local system stored database which contains the data for analysis.

- Using sqlalchemy based string for connecting the database locally stored in MySQL database management system. 
- Locally stored database InternationalDebtAnalysis has International_Debt_Data table which we will be using in this notebook for analysis.

In [5]:
# The Connection String for MySQL Databases.
%sql mysql+mysqldb://root:transformer@localhost/internationaldebtanalysis

In [43]:
# The first look of the data .
%%sql 
select * from international_debt_data 
LIMIT 10

 * mysql+mysqldb://root:***@localhost/internationaldebtanalysis
10 rows affected.


[('Afghanistan', 'AFG', 'DT.DIS.DLXF.CD', 'Disbursements on external debt', Decimal('833353922.900000000')),
 ('Afghanistan', 'AFG', 'DT.DIS.DPPG.CD', 'Disbursements on external debt', Decimal('833353922.900000000')),
 ('Afghanistan', 'AFG', 'DT.GPA.DPPG', 'Average grace period on new external debt commitments (years)', Decimal('66.249900000')),
 ('Afghanistan', 'AFG', 'DT.GRE.DPPG', 'Average grant element on new external debt commitments (%)', Decimal('247.400800000')),
 ('Afghanistan', 'AFG', 'DT.INR.DPPG', 'Average interest on new external debt commitments (%)', Decimal('6.884300000')),
 ('Afghanistan', 'AFG', 'DT.MAT.DPPG', 'Average maturity on new external debt commitments (years)', Decimal('176.416500000')),
 ('Afghanistan', 'AFG', 'DT.DOD.ALLC.ZS', 'Concessional debt (% of total external debt)', Decimal('461.567200000')),
 ('Afghanistan', 'AFG', 'BN.CAB.XOKA.CD', 'Current account balance (current US$)', Decimal('-11907343860.600000000')),
 ('Afghanistan', 'AFG', 'DT.DSB.DPPG.CD', 'Debt buyback (current US$)', Decimal('0E-9')),
 ('Afghanistan', 'AFG', 'DT.DOD.MDRI.CD', 'Debt forgiveness grants (current US$)', Decimal('165630000.000000000'))]

# . Finding the number of distinct countries.

From the first ten rows, we can see the amount of debt owed by Afghanistan in the different debt indicators. But we do not know the number of different countries we have on the table. There are repetitions in the country names because a country is most likely to have debt in more than one debt indicator.

Without a count of unique countries, we will not be able to perform our statistical analyses holistically. In this section, we are going to extract the number of unique countries present in the table.

In [44]:
%%sql 
SELECT COUNT(DISTINCT Country_Name) as total_different_countries 
from international_debt_data;

 * mysql+mysqldb://root:***@localhost/internationaldebtanalysis
1 rows affected.


[(122,)]

# . Finding out the distinct debt indicators.

We can see there are a total of 122 countries present on the table. As we saw in the first section, there is a column called indicator_name that briefly specifies the purpose of taking the debt. Just beside that column, there is another column called indicator_code which symbolizes the category of these debts. Knowing about these various debt indicators will help us to understand the areas in which a country can possibly be indebted to.

In [27]:
%%sql 
SELECT DISTINCT Indicator_Code as distinct_indiactor_codes , Indicator_Name
from international_debt_data;

 * mysql+mysqldb://root:***@localhost/internationaldebtanalysis
32 rows affected.


[('DT.DIS.DLXF.CD', 'Disbursements on external debt'),
 ('DT.DIS.DPPG.CD', 'Disbursements on external debt'),
 ('DT.GPA.DPPG', 'Average grace period on new external debt commitments (years)'),
 ('DT.GRE.DPPG', 'Average grant element on new external debt commitments (%)'),
 ('DT.INR.DPPG', 'Average interest on new external debt commitments (%)'),
 ('DT.MAT.DPPG', 'Average maturity on new external debt commitments (years)'),
 ('DT.DOD.ALLC.ZS', 'Concessional debt (% of total external debt)'),
 ('BN.CAB.XOKA.CD', 'Current account balance (current US$)'),
 ('DT.DSB.DPPG.CD', 'Debt buyback (current US$)'),
 ('DT.DOD.MDRI.CD', 'Debt forgiveness grants (current US$)'),
 ('DT.DFR.DPPG.CD', 'Debt forgiveness or reduction (current US$)'),
 ('DT.DSF.DPPG.CD', 'Debt stock reduction (current US$)'),
 ('DT.DXR.DPPG.CD', 'Debt stock rescheduled (current US$)'),
 ('DT.DOD.DECT.GN.ZS', 'External debt stocks (% of GNI)'),
 ('NY.GNP.MKTP.CD', 'GNI (current US$)'),
 ('DT.DIS.IDAG.CD', 'IDA grants (current US$)'),
 ('DT.IXF.DPPG.CD', 'Interest forgiven (current US$)'),
 ('DT.INT.DECT.GN.ZS', 'Interest payments on external debt (% of GNI)'),
 ('DT.IXR.DPPG.CD', 'Interest rescheduled (capitalized) (current US$)'),
 ('DT.DOD.MLAT.ZS', 'Multilateral debt (% of total external debt)'),
 ('DT.IXA.DPPG.CD.CG', 'Net change in interest arrears (current US$)'),
 ('DT.DOD.PVLX.GN.ZS', 'Present value of external debt (% of GNI)'),
 ('DT.DOD.PVLX.CD', 'Present value of external debt (current US$)'),
 ('BX.KLT.DREM.CD.DT', 'Primary income on FDI (current US$)'),
 ('DT.AXF.DPPG.CD', 'Principal forgiven (current US$)'),
 ('DT.AXR.DPPG.CD', 'Principal rescheduled (current US$)'),
 ('DT.DOD.DSTC.ZS', 'Short-term debt (% of total external debt)'),
 ('BX.GRT.TECH.CD.DT', 'Technical cooperation grants (current US$)'),
 ('DT.TXR.DPPG.CD', 'Total amount of debt rescheduled (current US$)'),
 ('DT.DOD.DECT.CD.CG', 'Total change in external debt stocks (current US$)'),
 ('FI.RES.TOTL.DT.ZS', 'Total reserves (% of total external debt)'),
 ('FI.RES.TOTL.MO', 'Total reserves in months of imports')]

# . Total amount of debt owed by the countries of the world.

As mentioned earlier, the financial debt of a particular country represents its economic state. But if we were to project this on an overall global scale, how will we approach it?

Let's switch gears from the debt indicators now and find out the total amount of debt (in USD) that is owed by the different countries. This will give us a sense of how the overall economy of the entire world is holding up.

In [16]:
%%sql 
SELECT ROUND((SUM(COALESCE(Debt)/100000)),2)
FROM international_debt_data;

 * mysql+mysqldb://root:***@localhost/internationaldebtanalysis
1 rows affected.


[(Decimal('3670503278.19'),)]

# . Top 5 Countries with highest debt. 

Now that we have the exact total of the amounts of debt owed by several countries, let's now find out the countries that owns the highest amount of debt along with the amount. Note that this debt is the sum of different debts owed by a countries across several categories. This will help to understand more about the countries in terms of its socio-economic scenarios. We can also find out the category in which the countries owns its highest debt. But we will leave that for some other analytical project.

In [20]:
%%sql 
SELECT Country_Name , SUM(Debt) as total_debt 
from international_debt_data
Group By Country_Name 
Order By total_debt desc
Limit 5 ;

 * mysql+mysqldb://root:***@localhost/internationaldebtanalysis
5 rows affected.


[('China', Decimal('92133233183022.388613306')),
 ('South Asia', Decimal('33614214683429.009656758')),
 ('Brazil', Decimal('32207558263344.782888521')),
 ('India', Decimal('26180650918494.559634648')),
 ('Russian Federation', Decimal('25408528988186.694150811'))]

Here we see that China owns the highest amount of debt which has very significant difference from the second position of South Asian countries.

# . Average amount of debt across different debt indicators.

We now have a brief overview of the dataset and a few of its summary statistics. We already have an idea of the different debt indicators in which the countries owe their debts.
      We can dig even further to find out on an average how much debt a country owes fro each debt indiactors ? This will give us a better sense of the distribution of the amount of debt across different indicators.

In [42]:
%%sql 
SELECT Indicator_Code as Debt_Indicator ,
Indicator_Name ,
AVG(Debt) as avg_debt 
FROM international_debt_data 
Group By 1,2
Order By 3 DESC;

 * mysql+mysqldb://root:***@localhost/internationaldebtanalysis
32 rows affected.


[('NY.GNP.MKTP.CD', 'GNI (current US$)', Decimal('2886728622305.7743902439024')),
 ('DT.DOD.DECT.CD.CG', 'Total change in external debt stocks (current US$)', Decimal('48990245907.7341463414634')),
 ('BX.KLT.DREM.CD.DT', 'Primary income on FDI (current US$)', Decimal('31400544875.8829958333333')),
 ('DT.TXR.DPPG.CD', 'Total amount of debt rescheduled (current US$)', Decimal('5527120251.5016260162602')),
 ('DT.DSF.DPPG.CD', 'Debt stock reduction (current US$)', Decimal('3724538279.5609756097561')),
 ('BX.GRT.TECH.CD.DT', 'Technical cooperation grants (current US$)', Decimal('3197601056.9105691056911')),
 ('DT.DOD.MDRI.CD', 'Debt forgiveness grants (current US$)', Decimal('2807907154.4715447154472')),
 ('DT.DXR.DPPG.CD', 'Debt stock rescheduled (current US$)', Decimal('2149461034.1528455284553')),
 ('DT.AXR.DPPG.CD', 'Principal rescheduled (current US$)', Decimal('2080190134.8878048780488')),
 ('DT.AXF.DPPG.CD', 'Principal forgiven (current US$)', Decimal('1423106269.5853658536585')),
 ('DT.IXR.DPPG.CD', 'Interest rescheduled (capitalized) (current US$)', Decimal('976847436.5479674796748')),
 ('DT.DIS.DLXF.CD', 'Disbursements on external debt', Decimal('833353922.9000000000000')),
 ('DT.DIS.DPPG.CD', 'Disbursements on external debt', Decimal('833353922.9000000000000')),
 ('DT.DSB.DPPG.CD', 'Debt buyback (current US$)', Decimal('686841406.4447154471545')),
 ('DT.IXF.DPPG.CD', 'Interest forgiven (current US$)', Decimal('382045591.6447154471545')),
 ('DT.DIS.IDAG.CD', 'IDA grants (current US$)', Decimal('361104979.2330081300813')),
 ('DT.IXA.DPPG.CD.CG', 'Net change in interest arrears (current US$)', Decimal('14624878.1772357723577')),
 ('DT.DOD.DECT.GN.ZS', 'External debt stocks (% of GNI)', Decimal('2168.8425286640081')),
 ('FI.RES.TOTL.DT.ZS', 'Total reserves (% of total external debt)', Decimal('1899.5702077555366')),
 ('DT.DOD.MLAT.ZS', 'Multilateral debt (% of total external debt)', Decimal('1176.5262406120244')),
 ('DT.DOD.ALLC.ZS', 'Concessional debt (% of total external debt)', Decimal('1098.1516126272358')),
 ('DT.MAT.DPPG', 'Average maturity on new external debt commitments (years)', Decimal('867.4298702629268')),
 ('DT.GRE.DPPG', 'Average grant element on new external debt commitments (%)', Decimal('679.0766176042358')),
 ('DT.DOD.DSTC.ZS', 'Short-term debt (% of total external debt)', Decimal('445.8459428474228')),
 ('DT.GPA.DPPG', 'Average grace period on new external debt commitments (years)', Decimal('248.1509040239593')),
 ('DT.INR.DPPG', 'Average interest on new external debt commitments (%)', Decimal('140.3106650595528')),
 ('FI.RES.TOTL.MO', 'Total reserves in months of imports', Decimal('109.1317763603171')),
 ('DT.INT.DECT.GN.ZS', 'Interest payments on external debt (% of GNI)', Decimal('57.0824771276829')),
 ('DT.DOD.PVLX.GN.ZS', 'Present value of external debt (% of GNI)', Decimal('0E-13')),
 ('DT.DOD.PVLX.CD', 'Present value of external debt (current US$)', Decimal('0E-13')),
 ('BN.CAB.XOKA.CD', 'Current account balance (current US$)', Decimal('-1854405642.3675203256911')),
 ('DT.DFR.DPPG.CD', 'Debt forgiveness or reduction (current US$)', Decimal('-4460948621.2471544715447'))]

By World Bank Indicator meta data analysis it is came into information that GNI stands for Gross National Income and it 
can not be considered as a debt indicator or debt category it is moreover the information about countries GNI which is later 
considered for another important debt indicator like (GNI/Debt percent ) .

So we are going to consider the indicator with seocnd most avgerage debt and that is DT.DOD.DECT.CD.CG .
      According to the World Bank metadata dataset this Indicators tells us about the external debt owed to nonresidents repayable in currency, goods, or services. 
      Total external debt is the sum of public, publicly guaranteed, and private nonguaranteed long-term debt, use of IMF credit, and short-term debt.

# . Country with highest amount of Total External Debt owed to Nonresidents.

We can see that the indicator DT.DOD.DECT.CD.CG comes at second position after GNI (not a debt indicator precisely) in the chart of average debt. This category includes the external debt owed to nonresidents repayable in currency, goods, or services. 

We can investigate this a bit more so as to find out which country owes the highest amount of debt in the category of (DT.DOD.DECT.CD.CG) Total External Debt owed to Nonresidents. Since not all the countries suffer from the same kind of economic disturbances, this finding will allow us to understand that particular country's economic condition a bit more specifically.

In [40]:
%%sql
SELECT 
    Country_Name, 
    Indicator_Name
FROM international_debt_data
WHERE Debt= (SELECT 
                 MAX(Debt)
             FROM international_debt_data
             WHERE indicator_code ='DT.DOD.DECT.CD.CG');

 * mysql+mysqldb://root:***@localhost/internationaldebtanalysis
1 rows affected.


[('China', 'Total change in external debt stocks (current US$)')]

So it is CHINA that has highest debt in this category which has highest average debt in all the categories.

# . The most common debt indiactors.

Lets find out the most common indicators in whihch contries owe their debts. 

In [49]:
%%sql
SELECT 
    Indicator_Code,
    COUNT(Indicator_Code) AS Indicator_Count
FROM International_Debt_Data
GROUP BY Indicator_Code
ORDER BY Indicator_Count DESC, Indicator_Code DESC
Limit 10;

 * mysql+mysqldb://root:***@localhost/internationaldebtanalysis
10 rows affected.


[('NY.GNP.MKTP.CD', 123),
 ('FI.RES.TOTL.MO', 123),
 ('FI.RES.TOTL.DT.ZS', 123),
 ('DT.TXR.DPPG.CD', 123),
 ('DT.MAT.DPPG', 123),
 ('DT.IXR.DPPG.CD', 123),
 ('DT.IXF.DPPG.CD', 123),
 ('DT.IXA.DPPG.CD.CG', 123),
 ('DT.INT.DECT.GN.ZS', 123),
 ('DT.INR.DPPG', 123)]

# . Some More Statistics Around Debt.

Let's change tracks from debt_indicators now and focus on the amount of debt again. Let's find out the maximum amount of debt across the indicators along with the respective country names. With this, we will be in a position to identify the other plausible economic issues a country might be going through. By the end of this section, we will have found out the debt indicators in which a country owes its highest debt.

In this notebook, we took a look at debt owed by countries across the globe. We extracted a few summary statistics from the data and unraveled some interesting facts and figures. We also validated our findings to make sure the investigations are correct.

In [47]:
%%sql
SELECT 
    Country_Name, 
    Indicator_Code, 
    MAX(Debt) AS Maximum_Debt
FROM International_Debt_Data
GROUP BY 1,2
ORDER BY 3 DESC
LIMIT 10;

 * mysql+mysqldb://root:***@localhost/internationaldebtanalysis
10 rows affected.


[('China', 'NY.GNP.MKTP.CD', Decimal('87868052932082.000000000')),
 ('South Asia', 'NY.GNP.MKTP.CD', Decimal('33359009781335.000000000')),
 ('Brazil', 'NY.GNP.MKTP.CD', Decimal('31975509879449.000000000')),
 ('India', 'NY.GNP.MKTP.CD', Decimal('26031108390207.000000000')),
 ('Russian Federation', 'NY.GNP.MKTP.CD', Decimal('23142116000000.000000000')),
 ('Mexico', 'NY.GNP.MKTP.CD', Decimal('22187452195572.000000000')),
 ('Turkiye', 'NY.GNP.MKTP.CD', Decimal('12517108682381.000000000')),
 ('Least developed countries: UN classification', 'NY.GNP.MKTP.CD', Decimal('10986971398367.000000000')),
 ('Indonesia', 'NY.GNP.MKTP.CD', Decimal('10527300786212.000000000')),
 ('Argentina', 'NY.GNP.MKTP.CD', Decimal('9032337724558.000000000'))]

 Here we see that NY.GNP.MKTP.CD is the indicator code in which most of the contries has their highest debt . Lets see what it is ?
 So according to World Bank Meta Data Glossary NY.GNP.MKTP.CD stands for the Total external debt stocks to gross national income. Total external debt is debt owed to nonresidents repayable in currency, goods, or services. Total external debt is the sum of public, publicly guaranteed, and private nonguaranteed long-term debt, use of IMF credit, and short-term debt.
 Short-term debt includes all debt having an original maturity of one year or less and interest in arrears on long-term debt.  
 GNI (formerly GNP) is the sum of value added by all resident producers plus any product taxes (less subsidies) not included in the valuation of output plus net receipts of primary income (compensation of employees and property income) from abroad.
        


# . Conclusion

In this notebook we came to know about the differnet aspects of debt which countries owes from the World Bank International Debt Data and made us more aware about the financial situations of the countries. This analysis of debt is done on more higher levels by economists to get more interesting and accurate results on the economies of different countries of the world.


So to end with this analytical project on debt analysis we can say on the lighter note that we all owe this world a lot !!